### PR TITLE
make required captcha score configurable

### PR DESCRIPTION
**Pull request by mfrancisc**
Use new configuration options:
- `toolchainconfig.Spec.Host.RegistrationService.Verification.Captcha.RequiredScore` which blocks usersginup below this value (if it's not a reactivation)
- `toolchainconfig.Spec.Host.RegistrationService.Verification.Captcha. AllowLowScoreReactivation` is enabled by default and allows reactivations to proceed even when captcha score is below the configured `RequiredScore`.

Related PR:
- toolchain-common: https://github.com/codeready-toolchain/toolchain-common/pull/344
- sandbox-sre: https://github.com/codeready-toolchain/sandbox-sre/pull/1348

Jira: https://issues.redhat.com/browse/SANDBOX-495

### CODE REVIEW

File: pkg/verification/service/verification_service_test.go
Comment (alexeykazakov) [inline pkg/verification/service/verification_service_test.go:1171]: Alternatively you could organize the tests as an array of cases where you provide:
- the score and activation counter annotations
- configuration
- expected result
And then go over them to test. It will reduce code duplication and will be easy to add more cases.

Could you please also add tests to cover cases with:
- there is no score annotation at all.
- score annotation is not valid

**Review by alexeykazakov — COMMENTED** [review]
Looks great! Just a few minor comments.

File: pkg/verification/service/verification_service.go
Comment (mfrancisc) [reply pkg/verification/service/verification_service.go:247]: thanks!

File: pkg/verification/service/verification_service.go
Comment (mfrancisc) [reply pkg/verification/service/verification_service.go:361]: yeah, better to parse the score only when found. Good catch!

File: pkg/verification/service/verification_service_test.go
Comment (mfrancisc) [reply pkg/verification/service/verification_service_test.go]: > Could you please also add tests to cover cases 

Thanks! Done in the same commit. But I'm not 100% sure what should happen in those cases, seems like currently the verification process just goes on, ignoring that the captcha score is missing/invalid. So I just did that. 

Not sure if we should log/return an error or if this is already being done in other parts of the verification process, I guess before going to the phone verification process.

Comment (alexeykazakov) [reply pkg/verification/service/verification_service_test.go]: If the score is missing that we should treat it as a high score. For invalid scores ideally we would want to return an error but it's OK if we just log the error and ignore it for now.

File: pkg/configuration/configuration.go
Comment (mfrancisc) [reply pkg/configuration/configuration.go:286]: Right now it is already using the hard-coded `0.6` value : https://github.com/codeready-toolchain/registration-service/blob/ec03997ea2af344078922dec3dddd65a2a0435bc/pkg/verification/service/verification_service.go#L234
We can set it to 0, but this means we have to open and merge a PR in sandbox-sre with the `0.6` threshold before merging this. Not sure if it makes any difference 🤷‍♂️ . Any preference ?

File: pkg/verification/service/verification_service.go
Comment (rajivnathan) [inline pkg/verification/service/verification_service.go:251]: I may be missing something but it looks like all cases lead to calling `checkRequiredCaptchaScore(ctx, signup, cfg)`. If CaptchaAllowLowScoreReactivation is true and it's a reactivation then doesn't that mean we don't need to do additional (phone) verification and we can just proceed to activate the user?

So maybe we can simplify the logic by removing the CaptchaAllowLowScoreReactivation check from here and move it to the IsPhoneVerificationRequired function in https://github.com/codeready-toolchain/registration-service/blob/1602277a54b638f1056f52ef77757f41cc61d2d5/pkg/signup/service/signup_service.go#L227C2-L232 so that we skip phone verification at that point instead and keep the logic simple here?

**Review by rajivnathan — COMMENTED** [review]
I was getting a bit confused by the logic around `checkRequiredCaptchaScore`. There are 3 segments for scores right?

1. `UserScore` > `ScoreThreshold` (no additional verification required)
2. `ScoreThreshold` > `UserScore` > `RequiredScore` (additional verification required)
3. `RequiredScore` > `UserScore` (manual verification required)

Just an idea but maybe it would be easier to understand the logic if `checkRequiredCaptchaScore` is renamed to `checkRequiredManualApproval`?

When zooming out of the scope of this specific PR I think it's easy to confuse `ScoreThreshold` and `RequiredScore`. Something like `AdditionalVerificationThreshold` maybe would have been a bit more clear.

File: pkg/verification/service/verification_service.go
Comment (mfrancisc) [reply pkg/verification/service/verification_service.go:251]: > I may be missing something but it looks like all cases lead to calling checkRequiredCaptchaScore(ctx, signup, cfg).

the case in which is `activationCounter > 1` and `CaptchaAllowLowScoreReactivation == true` doesn't lead to `checkRequiredCaptchaScore` . Maybe I can shuffle the if statements a bit to make this more clear 🤔

**Comment by mfrancisc** [comment]
> Just an idea but maybe it would be easier to understand the logic if checkRequiredCaptchaScore is renamed to checkRequiredManualApproval?

I like the name more , thanks for the feedback. Renamed in https://github.com/codeready-toolchain/registration-service/pull/379/commits/d5a6f9ba04cc8e77865bdc3387151ff6f83e7efe

> When zooming out of the scope of this specific PR I think it's easy to confuse ScoreThreshold and RequiredScore. Something like AdditionalVerificationThreshold maybe would have been a bit more clear. And something like CaptchaVerificationThreshold probably would have been better than ScoreThreshold.

This makes sense indeed. As all those captcha configuration threshold can be a bit confusing now. But let's do this in separate PRs , WDYT ?

File: pkg/verification/service/verification_service.go
Comment (rajivnathan) [reply pkg/verification/service/verification_service.go]: >Not sure, we could move it there but we need to pass the activation counter from the user. Do you think that would make it simpler ?

You're right, that part of the code doesn't reference the UserSignup so I guess it's more complicated. My bad, disregard the suggestion.

File: pkg/configuration/configuration.go
Comment (rajivnathan) [reply pkg/configuration/configuration.go:286]: I think 0 as a default makes more sense too, that way it's permissible by default and you only need to set it if you want to limit automatic approval.

Comment (mfrancisc) [reply pkg/configuration/configuration.go]: makes sense. I've changed the default to `0` in https://github.com/codeready-toolchain/registration-service/pull/379/commits/a7c22771f1d488dbc1525a6e7018b26ac05261ab. Thanks!

Comment (mfrancisc) [reply pkg/configuration/configuration.go]: and related sandbox-sre PR: https://github.com/codeready-toolchain/sandbox-sre/pull/1348

File: pkg/verification/service/verification_service_test.go
Comment (alexeykazakov) [inline pkg/verification/service/verification_service_test.go:903]: Shouldn't it be:
```suggestion
			allowLowScoreReactivationConfiguration       bool
```
?

File: pkg/verification/service/verification_service_test.go
Comment (alexeykazakov) [inline pkg/verification/service/verification_service_test.go:928]:
```suggestion
			"allow low score reactivation disabled - captcha score below required score and it's a reactivation": {
```

File: pkg/verification/service/verification_service_test.go
Comment (alexeykazakov) [inline pkg/verification/service/verification_service_test.go:934]:
```suggestion
			"allow low score reactivation disabled - captcha score below required score and it's not a reactivation": {
```

File: pkg/verification/service/verification_service_test.go
Comment (alexeykazakov) [inline pkg/verification/service/verification_service_test.go:940]:
```suggestion
			"allow low score reactivation disabled - captcha score ok": {
```

File: pkg/verification/service/verification_service_test.go
Comment (alexeykazakov) [inline pkg/verification/service/verification_service_test.go:945]: Let's also add a test for the case with no activation counter annotation set at all.

**Review by alexeykazakov — APPROVED** [review]
Thank you for addressing the comments! The test look much better now! I left a few minor comments. But overall looks good!

File: pkg/verification/service/verification_service_test.go
Comment (mfrancisc) [reply pkg/verification/service/verification_service_test.go:945]: good point! done , please see https://github.com/codeready-toolchain/registration-service/pull/379/commits/c09beb3d59e404b08a99285b2d7e6437a5b4b3df

File: pkg/verification/service/verification_service.go
Comment (MatousJobanek) [reply pkg/verification/service/verification_service.go]: never mind - I made the comment before I saw that you didn't use `0` is the default value, so I understand why it was confusing.
for the background of my question - I was just wondering if we need to do all these checks when the value is `0`, couldn't we just skip it? However, I understand that logical-wise it doesn't change anything.

**Review by MatousJobanek — APPROVED** [review]
:+1:

File: pkg/verification/service/verification_service.go
Comment (mfrancisc) [reply pkg/verification/service/verification_service.go]: ok, got it! thanks for explaining that and for the review 👍

**Review by rajivnathan — APPROVED** [review]
Nice 👍

**Pull request merged**
